README.md
telly
===============

The telly project is a script designed to relieve the manual effort required when setting TestRail results for automated Beaker tests

**The telly project is very much in Alpha right now, and not recommended for use**

Download
--

Soon a gem will be put on rubygems.org

For now you can package and install the gem manually with 

`gem build telly.gemspec`

`gem install telly-0.1.0.gem`


Usage
--
`telly -t 1234123121 -j beaker_tests/junit/latest/beaker_junit.xml`

Currently using the script requires a few things.

* A Beaker junit output file
* The Test Run ID from a testrail test run. This can be found in the upper left of the test run page
* Your beaker scripts must include the Test Case ID in the test_name
    * e.g. `test_name "QENG-694 - C57870 - This is a well written test name"`
    * There is a plan to remove this requirement, see the Roadmap below

The script displays output showing progress, and on completion will print any test cases that failed to submit their results to TestRail. These might be incorrect test case IDs or authentication problems. Currently there isn't a good way to retry a submission and only submit the previous failures. It simply re-submits all of the beaker results. 

Roadmap
--

Since this project is currently just a proof of concept, there is a lot that has to be done.

To make it generally usable for others, the plan is:

* Create spec tests (in progress)
* Harden the code
    * be able to properly handle failures to post individual data
    * be able to properly handle net failure part way through posting
    * error handling for bad user provided data, failures
* Release gem for testing
    * Code is all packaged up to be made into a gem, but is not on rubygems yet

After this initial release, we can potentially expand this to integrate directly with the TestRail website and jenkins to execute test runs with the click of a button and have the results fed back into testrail automatically.

TODO
--
**Automated Test Links**

In order to remove the requirement of having testcase IDs in beaker scripts, the QA team has suggested using the Automated Test Link field from TestRail to match up beaker test output to a TestRail test case.

This would involve using the [get_tests](http://docs.gurock.com/testrail-api2/reference-tests#get_tests) api call to get the tests from a test run, examining the Automated Test Link field, and matching the github links to the beaker test scripts in the junit output.

It would be harder to implement but much simpler for the user, since putting the testcase ID in the beaker script isn't universal among the QA teams

Useful Links
--
If you are hacking on this project, the [TestRail API docs](http://docs.gurock.com/testrail-api2/start) might come in handy


HISTORY.md
# default - History
## Tags
* [LATEST - 1 Jul, 2015 (3dfb6906)](#LATEST)
* [0.1.2 - 17 Feb, 2015 (7ed8f7e1)](#0.1.2)
* [0.1.1 - 26 Jan, 2015 (b88029bd)](#0.1.1)

## Details
### <a name = "LATEST">LATEST - 1 Jul, 2015 (3dfb6906)

* (GEM) update telly version to 0.2.0 (3dfb6906)

* Merge pull request #9 from anodelman/gem (afa8ab0c)


```
Merge pull request #9 from anodelman/gem

(MAINT) bad version number, accidental copy-paste from another project
```
* (MAINT) bad version number, accidental copy-paste from another project (0e3c092d)


```
(MAINT) bad version number, accidental copy-paste from another project

- should be at current released version 0.1.2
```
* (QENG-2616) Telly - Cannot Find Test Files Using Beaker... (55c17375)


```
(QENG-2616) Telly - Cannot Find Test Files Using Beaker...

... 2.8+ Logs/JUnit Reports

- beaker added a directory level that telly couldn't handle
- addeds support for --dry-run, which just process the junit file
  without creating a TestRail api connection or attempting to submit
  results
```
* (QENG-2195) add telly jenkins release pipeline (5d0ef508)


```
(QENG-2195) add telly jenkins release pipeline

- add spec test rake task
```
* (QENG-2195) add telly jenkins release pipeline (f74a0c79)


```
(QENG-2195) add telly jenkins release pipeline

- standardized gem workflow
```
### <a name = "0.1.2">0.1.2 - 17 Feb, 2015 (7ed8f7e1)

* Merge pull request #5 from colinPL/maint_v0.1.2_gem (7ed8f7e1)


```
Merge pull request #5 from colinPL/maint_v0.1.2_gem

(GEM) Update telly version to 0.1.2
```
* (GEM) Update telly version to 0.1.2 (8bb2f561)

* Merge pull request #4 from jpinsonault/QENG-1796-improve-error-recovery (2d825896)


```
Merge pull request #4 from jpinsonault/QENG-1796-improve-error-recovery

(QENG-1796) Improved error recovery
```
* Created custom exception, better rescuing (e23c82c0)

* (QENG-1796) Improved error recovery (a3161911)


```
(QENG-1796) Improved error recovery
Previously only testrail api errors were caught and shown to the
user, now any exception that happens during submission is caught
and that test is reported to the user.

In addition, when a beaker test file doesn't contain a testcase id
it raises an exception with an appropriate error.

This PR was motivated by pre-suites not having testcase ids
causing the script to throw an uncaught exception
```
### <a name = "0.1.1">0.1.1 - 26 Jan, 2015 (b88029bd)

* Initial release.


spec/spec_helper.rb
require 'webmock/rspec'
require 'telly'


spec/testrail_spec.rb
require_relative "../lib/telly"
require_relative "../lib/testrail"
require_relative './spec_helper'


describe Telly do
  class BTModule
  end

  before :each do
    @testrail_credentials = {
      "testrail_username" => "test",
      "testrail_password" => "testpass"
    }
  end

  describe "#get_testrail_api" do
    it "returns a APIClient object" do
      api = Telly::get_testrail_api(@testrail_credentials)
      expect(api).to be_an_instance_of TestRail::APIClient
    end

    it "has the correct credentials" do
      api = Telly::get_testrail_api(@testrail_credentials)
      expect(api.password).to be @testrail_credentials["testrail_password"]
      expect(api.user).to be @testrail_credentials["testrail_username"]
    end
  end
end


lib/telly/test_rail.rb
#
# TestRail API binding for Ruby (API v2, available since TestRail 3.0)
#
# Learn more:
#
# http://docs.gurock.com/testrail-api2/start
# http://docs.gurock.com/testrail-api2/accessing
#
# Copyright Gurock Software GmbH. See license.md for details.
#

require 'net/http'
require 'net/https'
require 'uri'
require 'json'

module Telly
  module TestRail
    class APIClient
      @url = ''
      @user = ''
      @password = ''

      attr_accessor :user
      attr_accessor :password

      def initialize(base_url)
        if !base_url.match(/\/$/)
          base_url += '/'
        end
        @url = base_url + 'index.php?/api/v2/'
      end

      #
      # Send Get
      #
      # Issues a GET request (read) against the API and returns the result
      # (as Ruby hash).
      #
      # Arguments:
      #
      # uri                 The API method to call including parameters
      #                     (e.g. get_case/1)
      #
      def send_get(uri) 
        _send_request('GET', uri, nil)
      end

      #
      # Send POST
      #
      # Issues a POST request (write) against the API and returns the result
      # (as Ruby hash).
      #
      # Arguments:
      #
      # uri                 The API method to call including parameters
      #                     (e.g. add_case/1)
      # data                The data to submit as part of the request (as
      #                     Ruby hash, strings must be UTF-8 encoded)
      #
      def send_post(uri, data)
        _send_request('POST', uri, data)
      end

      private
      def _send_request(method, uri, data)
        url = URI.parse(@url + uri)
        if method == 'POST'
          request = Net::HTTP::Post.new(url.path + '?' + url.query)
          request.body = JSON.dump(data)
        else
          request = Net::HTTP::Get.new(url.path + '?' + url.query)
        end
        request.basic_auth(@user, @password)
        request.add_field('Content-Type', 'application/json')

        conn = Net::HTTP.new(url.host, url.port)
        if url.scheme == 'https'
          conn.use_ssl = true
          conn.verify_mode = OpenSSL::SSL::VERIFY_NONE
        end
        response = conn.request(request)

        if response.body && !response.body.empty?
          result = JSON.parse(response.body)
        else
          result = {}
        end

        if response.code != '200'
          if result && result.key?('error')
            error = '"' + result['error'] + '"'
          else
            error = 'No additional error message received'
          end
          raise APIError.new('TestRail API returned HTTP %s (%s)' %
            [response.code, error])
        end
        
        result
      end
    end

    class APIError < StandardError
    end

  end
end


lib/beaker_testrail.rb
#!/usr/bin/env ruby

require 'nokogiri'
require 'yaml'
require 'pp'
require 'testrail'

# == beaker_testrail.rb
# This module provides functions to add test results in testrail from a 
# finished beaker run. 
# 
# It takes in a beaker junit file and a TestRail Testrun ID
# 
# It matches the beaker tests with TestRail testcases by looking for the 
#   test case ID in the beaker script. The combination of a test run and a test case
#   allows the script to add a result for a particular instance of a test case.
#   In TestRail parlance, this is confusingly called a test.
# From the TestRail API docs:
#   "In TestRail, tests are part of a test run and the test cases are part of the
#   related test suite. So, when you create a new test run, TestRail creates a test
#   for each test case found in the test suite of the run. You can therefore think
#   of a test as an “instance” of a test case which can have test results, comments 
#   and a test status.""
module BeakerTestrail

  TESTRAIL_URL = 'https://testrail.ops.puppetlabs.net/'
  CREDENTIALS_FILE = '~/.testrail_credentials.yaml'

  # Used for extracted the test case ID from beaker scripts
  TESTCASE_ID_REGEX = /.*(?<jira_ticket>\w+-\d+).*[cC](?<testrun_id>\d+)/

  # Testrail Status IDs
  PASSED = 1
  BLOCKED = 2
  FAILED = 5

  def do_stub_test(credentials)
    api = get_testrail_api(credentials)

    api.send_post
  end


  ##################################
  # Main
  ##################################

  # Run the importer
  # 
  # @param [Hash] An optparse object
  #
  # @return [Void]
  #
  # @example password = BeakerTestrail::main(parse_opts)
  def BeakerTestrail.main(options)
    # Get pass/fail/skips from junit file
    results = load_junit_results(options[:junit_file])

    puts "Run results:"
    puts "#{results[:passes].length} Passing"
    puts "#{results[:failures].length} Failing or Erroring"
    puts "#{results[:skips].length} Skipped"

    # Set results in testrail
    bad_results = set_testrail_results(results, options[:junit_file], options[:testrun_id])

    # Print error messages
    if not bad_results.empty?
      puts "Error: There were problems processing these test scripts:"
      bad_results.each do |test_script, error|
        puts "#{test_script}:\n\t#{error}"
      end
    end
  end


  ##################################
  # TestRail API
  ##################################

  # Load testrail credentials from file
  #
  # @return [Hash] Contains testrail_username and testrail_password
  #
  # @example password = load_credentials()["testrail_password"]
  def BeakerTestrail.load_credentials(credentials_file)
    begin
      YAML.load_file(File.expand_path(credentials_file))  
    rescue
      puts "Error: Could not find #{credentials_file}"
      puts "Create #{credentials_file} with the following:"
      puts "testrail_username: your.username\ntestrail_password: yourpassword"

      exit
      
    end
  end


  # Returns a testrail API object that talks to testrail
  # 
  # @param [Hash] credentials A hash containing at least two keys, testrail_username and testrail_password
  #
  # @return [TestRail::APIClient] The API object for talking to TestRail
  #
  # @example api = get_testrail_api(load_credentials)
  def BeakerTestrail.get_testrail_api(credentials)
    client = TestRail::APIClient.new(TESTRAIL_URL)
    client.user = credentials["testrail_username"]
    client.password = credentials["testrail_password"]

    return client
  end

  # Sets the results in testrail. 
  # Tests that have testrail API exceptions are kept track of in bad_results
  #
  # @param [Hash] results A hash of lists of xml objects from the junit output file.
  # @param [String] junit_file The path to the junit xml file
  #                 Needed for determining the path of the test file in add_failure, etc
  # @param [String] testrun_id The TestRail test run ID
  # 
  # @return [Void] 
  #
  def BeakerTestrail.set_testrail_results(results, junit_file, testrun_id)
    credentials = load_credentials(CREDENTIALS_FILE)
    api = get_testrail_api(credentials)

    # Results that couldn't be set in testrail for some reason
    bad_results = {}

    # passes
    results[:passes].each do |junit_result|
      begin
        submit_result(api, PASSED, junit_result, junit_file, testrun_id)    
      rescue TestRail::APIError => e
        bad_results[junit_result[:name]] = e.message
      end
    end

    # Failures
    results[:failures].each do |junit_result|
      begin
        submit_result(api, FAILED, junit_result, junit_file, testrun_id)    
      rescue TestRail::APIError => e
        bad_results[junit_result[:name]] = e.message
      end
    end

    # Skips
    results[:skips].each do |junit_result|
      begin
        submit_result(api, BLOCKED, junit_result, junit_file, testrun_id)    
      rescue TestRail::APIError => e
        bad_results[junit_result[:name]] = e.message
      end
    end

    return bad_results
  end

  # Submits a test result to TestRail
  #
  # @param [TestRail::APIClient] api TestRail API object
  # @param [int] status The testrail status to set
  # @param [Nokogiri::XML::Element] junit_result The nokogiri node that holds the junit result
  # @param [String] junit_file Path to the junit file the test result originated from
  # @param [String] testrun_id The testrun ID
  #
  # @return [Void]
  # 
  # @raise [TestRail::APIError] When there is a problem with the API request, testrail raises
  #                             this exception. Should be caught for error reporting
  #
  # @example submit_result(api, BLOCKED, junit_result, junit_file, testrun_id)
  def BeakerTestrail.submit_result(api, status, junit_result, junit_file, testrun_id)
    test_file_path = beaker_test_path(junit_file, junit_result)

    puts junit_result.class
    testcase_id = testcase_id_from_beaker_script(test_file_path)

    time_elapsed = make_testrail_time(junit_result[:time])

    # Make appropriate comment for testrail
    case status
    when FAILED
      error_message = junit_result.xpath('./failure').first[:message]
      testrail_comment = "Failed with message:\n#{error_message}"
    when BLOCKED
      skip_message = junit_result.xpath('system-out').first.text
      testrail_comment = "Skipped with message:\n#{skip_message}"
    else
      testrail_comment = "Passed"
    end

    puts "\nSetting result for test case: #{testcase_id}"
    puts "Adding comment:\n#{testrail_comment}"

    api.send_post("add_result_for_case/#{testrun_id}/#{testcase_id}", 
      {
        status_id: status,
        comment: testrail_comment,
        elapsed: time_elapsed,
      }
    )
  end


  # Returns a string that testrail accepts as an elapsed time
  # Input from beaker is a float in seconds, so it rounds it to the 
  # nearest second, and adds an 's' at the end
  # 
  # Testrail throws an exception if it gets "0s", so it returns a 
  # minimum of "1s"
  #
  # @param [String] seconds_string A string that contains only a number, the elapsed time of a test
  #
  # @return [String] The elapsed time of the test run, rounded and with an 's' appended
  #
  # @example puts make_testrail_time("2.34") # "2s"
  def BeakerTestrail.make_testrail_time(seconds_string)
    # If time is 0, make it 1
    rounded_time = [seconds_string.to_f.round, 1].max
    # Test duration
    time_elapsed = "#{rounded_time}s"

    return time_elapsed
  end


  ##################################
  # Junit and Beaker file functions
  ##################################

  # Loads the results of a beaker run.
  # Returns hash of failures, passes, and skips that each hold a list of 
  # junit xml objects
  #
  # @param [String] junit_file Path to a junit xml file
  # 
  # @return [Hash] A hash containing xml objects for the failures, skips, and passes
  #
  # @example load_junit_results("~/junit/latest/beaker_junit.xml")
  def BeakerTestrail.load_junit_results(junit_file)
    junit_doc = Nokogiri::XML(File.read(junit_file))

    failures = junit_doc.xpath('//testcase[failure]')
    skips = junit_doc.xpath('//testcase[skip]')
    passes = junit_doc.xpath('//testcase[not(failure) and not(skip)]')

    return {failures: failures, skips: skips, passes: passes}
  end


  # Extracts the test case id from the test script
  #
  # @param [String] beaker_file Path to a beaker test script
  # 
  # @return [String] The test case ID
  #
  # @example testcase_id_from_beaker_script("~/tests/test_the_things.rb") # 1234
  def BeakerTestrail.testcase_id_from_beaker_script(beaker_file)
    # Find first matching line
    match = File.readlines(beaker_file).map { |line| line.match(TESTCASE_ID_REGEX) }.compact.first

    match[:testrun_id]
  end


  # Calculates the path to a beaker test file by combining the junit file path
  # with the test name from the junit results.
  # Makes the assumption that junit folder that beaker creates will always be 
  # 2 directories up from the beaker script base directory.
  # TODO somewhat hacky, maybe a config/command line option
  #
  # @param [String] junit_file_path Path to a junit xml file
  # @param [String] junit_result Path to a junit xml file
  # 
  # @return [String] The path to the beaker script from the junit test result
  #
  # @example load_junit_results("~/junit/latest/beaker_junit.xml")
  def BeakerTestrail.beaker_test_path(junit_file_path, junit_result)
    beaker_folder_path = junit_result[:classname]
    test_filename = junit_result[:name]

    File.join(File.dirname(junit_file_path), "../../", beaker_folder_path, test_filename)
  end

end


lib/telly/arg_parser.rb
require 'optparse'

module Telly
  class ArgParser

    VERSION_STRING =
"
 __o_____
()/O\\___()
 `-\\\\---' TELLY
              %s!
      __\\/__
     | .... |
     | .... |
      ------
"

    # Parses the command line options
    # @return [Void]
    def self.parse_opts
      options_hash = {}

      optparse = OptionParser.new do|parser|
        options_hash = {
          testrun_id: nil,
          junit_file: nil,
        }

        parser.on( '-t', '--testrun-id TESTRUN_ID', 'The testrun id' ) do |testrun_id|
          options_hash[:testrun_id] = testrun_id
        end

        parser.on( '-j', '--junit-folder JUNIT_FILE', 'Beaker junit file' ) do |junit_file|
          options_hash[:junit_file] = junit_file
        end

        parser.on( '-d', '--dry-run', 'Run without API connection to TestRail' ) do
          options_hash[:dry_run] = true
        end

        parser.on( '-h', '--help', 'Display this screen' ) do
          puts parser
          exit
        end

        parser.on( '-v', '--version', 'Report the current version number' ) do
          puts VERSION_STRING % Telly::Version::STRING
          exit
        end


        parser.parse!

        if not options_hash[:testrun_id] or not options_hash[:junit_file]
          puts "Error: Missing option(s)"
          puts parser
          exit
        end
      end

      options_hash
    end
  end
end


import_results.rb
#!/usr/bin/env ruby

# This script adds test results in testrail from a finished beaker run
# It takes in a beaker junit file and a TestRail Testrun ID
# 
# It matches the beaker tests with TestRail testcases by looking for the 
#   test case ID in the beaker script. The combination of a test run and a test case
#   allows the script to add a result for a particular instance of a test case.
#   In TestRail parlance, this is confusingly called a test.
# From the TestRail API docs:
#   "In TestRail, tests are part of a test run and the test cases are part of the
#   related test suite. So, when you create a new test run, TestRail creates a test
#   for each test case found in the test suite of the run. You can therefore think
#   of a test as an “instance” of a test case which can have test results, comments 
#   and a test status.""

require 'optparse'
require 'nokogiri'
require 'yaml'
require 'pp'
require_relative 'testrail'


TESTRAIL_URL = 'https://testrail.ops.puppetlabs.net/index.php'
CREDENTIALS_FILE = 'credentials.yaml'

# Used for extracted the test case ID from beaker scripts
TESTCASE_ID_REGEX = /.*(?<jira_ticket>\w+-\d+).*[cC](?<testrun_id>\d+)/

# Testrail Status IDs
PASSED = 1
BLOCKED = 2
FAILED = 5


##################################
# Main
##################################

def main(options)
  # Get pass/fail/skips from junit file
  results = load_junit_results(options[:junit_file])

  puts "Run results:"
  puts "#{results[:passes].length} Passing"
  puts "#{results[:failures].length} Failing or Erroring"
  puts "#{results[:skips].length} Skipped"

  # Set results in testrail
  set_testrail_results(results, options[:junit_file], options[:testrun_id])
end


# Parses the command line options
def parse_opts
  options_hash = {}

  optparse = OptionParser.new do|parser|
    options_hash = {
      testrun_id: nil,
      junit_file: nil,
    }

    parser.on( '-t', '--testrun-id TESTRUN_ID', 'The testrun id' ) do |testrun_id|
      options_hash[:testrun_id] = testrun_id
    end

    parser.on( '-j', '--junit-folder JUNIT_FILE', 'Beaker junit file' ) do |junit_file|
      options_hash[:junit_file] = junit_file
    end

    parser.on( '-h', '--help', 'Display this screen' ) do
      puts parser
      exit
    end

    parser.parse!

    if not options_hash[:testrun_id] or not options_hash[:junit_file]
      puts "Error: Missing option(s)"
      puts parser
      exit
    end
  end

  options_hash
end


##################################
# TestRail API
##################################

# Load testrail credentials from file
#
# ==== Returns
#
# +hash+ - Contains testrail_username and testrail_password
#
# ==== Examples
#
# password = load_credentials()["testrail_password"]
def load_credentials()
  begin
    YAML.load_file(CREDENTIALS_FILE)  
  rescue
    puts "Error: Could not find credentials.yaml\nHave you copied credentials_template.yaml and filled in your info?"

    exit
    
  end
end


# Returns a testrail API object that talks to testrail
#
# ==== Attributes
#
# * +credentials+ - A hash containing at least two keys, testrail_username and testrail_password
# 
# ==== Returns
#
# +TestRail::APIClient+ - The API object for talking to TestRail
#
# ==== Examples
#
# api = get_testrail_api(load_credentials)
def get_testrail_api(credentials)
  client = TestRail::APIClient.new(TESTRAIL_URL)
  client.user = credentials["testrail_username"]
  client.password = credentials["testrail_password"]

  return client
end

# Sets the results in testrail. 
# Tests that have testrail API exceptions are kept track of in bad_results
#
# ==== Attributes
#
# * +results+ - A hash of lists of xml objects from the junit output file.
# * +junit_file+ - The path to the junit xml file
#   Needed for determining the path of the test file in add_failure, etc
# * +testrun_id+ - The TestRail test run ID
# 
# ==== Returns
#
# nil
#
def set_testrail_results(results, junit_file, testrun_id)
  testrail_api = get_testrail_api(load_credentials)

  # Results that couldn't be set in testrail for some reason
  bad_results = {}

  # passes
  results[:passes].each do |junit_result|
    begin
      add_pass(testrail_api, junit_result, junit_file, testrun_id)    
    rescue TestRail::APIError => e
      bad_results[junit_result[:name]] = e.message
    end
  end

  # Failures
  results[:failures].each do |junit_result|
    begin
      add_failure(testrail_api, junit_result, junit_file, testrun_id)    
    rescue TestRail::APIError => e
      bad_results[junit_result[:name]] = e.message
    end
  end

  # Skips
  results[:skips].each do |junit_result|
    begin
      add_skip(testrail_api, junit_result, junit_file, testrun_id)    
    rescue TestRail::APIError => e
      bad_results[junit_result[:name]] = e.message
    end
  end

  # Print error messages
  if not bad_results.empty?
    puts "Error: There were problems processing these test scripts:"
    bad_results.each do |test_script, error|
      puts "#{test_script}:\n\t#{error}"
    end
  end
end


# Adds a fail result to a testcase
# Adds a comment to the result with the error message from the test run
def add_failure(testrail_api, junit_result, junit_file, testrun_id)
  test_file_path = beaker_test_path(junit_file, junit_result)

  testcase_id = testcase_id_from_beaker_script(test_file_path)

  time_elapsed = make_testrail_time(junit_result[:time])

  # Make an appropriate comment for the test's error message
  error_message = junit_result.xpath('./failure').first[:message]
  testrail_comment = "Failed with message:\n#{error_message}"

  puts "\nSetting result for failed test case: #{testcase_id}"
  puts "Adding comment:\n#{testrail_comment}"

  # TODO add user ID to result
  testrail_api.send_post("add_result_for_case/#{testrun_id}/#{testcase_id}", 
    {
      status_id: FAILED,
      comment: testrail_comment,
      elapsed: time_elapsed,
    }
  )
end


# Adds a pass result to a testcase
def add_pass(testrail_api, junit_result, junit_file, testrun_id)
  test_file_path = beaker_test_path(junit_file, junit_result)

  testcase_id = testcase_id_from_beaker_script(test_file_path)

  time_elapsed = make_testrail_time(junit_result[:time])

  puts "\nSetting result for passing test case: #{testcase_id}"

  testrail_api.send_post("add_result_for_case/#{testrun_id}/#{testcase_id}", 
    {
      status_id: PASSED,
      elapsed: time_elapsed
    }
  )
end


# Adds a pass result to a testcase
# Adds a comment with the skip message
def add_skip(testrail_api, junit_result, junit_file, testrun_id)
  test_file_path = beaker_test_path(junit_file, junit_result)

  testcase_id = testcase_id_from_beaker_script(test_file_path)

  time_elapsed = make_testrail_time(junit_result[:time])

  # Make an appropriate comment for the test's skip message
  skip_message = junit_result.xpath('system-out').first.text
  testrail_comment = "Skipped with message:\n#{skip_message}"

  puts "\nSetting result for skipped test case: #{testcase_id}"
  puts "Adding comment:\n#{testrail_comment}"

  testrail_api.send_post("add_result_for_case/#{testrun_id}/#{testcase_id}", 
    {
      status_id: BLOCKED,
      comment: testrail_comment,
      elapsed: time_elapsed
    }
  )
end


# Returns a string that testrail accepts as an elapsed time
# Input from beaker is a float in seconds, so it rounds it to the 
# nearest second, and adds an 's' at the end
# 
# Testrail throws an exception if it gets "0s", so it returns a 
# minimum of "1s"
#
# ==== Attributes
#
# * +seconds_string+ - A string that contains only a number, the elapsed time of a test
# 
# ==== Returns
#
# +string+ - The elapsed time of the test run, rounded and with an 's' appended
#
# ==== Examples
#
# puts make_testrail_time("2.34") # "2s"
def make_testrail_time(seconds_string)
  # If time is 0, make it 1
  rounded_time = [seconds_string.to_f.round, 1].max
  # Test duration
  time_elapsed = "#{rounded_time}s"

  return time_elapsed
end


##################################
# Junit and Beaker file functions
##################################

# Loads the results of a beaker run.
# Returns hash of failures, passes, and skips that each hold a list of 
# junit xml objects
#
# ==== Attributes
#
# * +junit_file+ - Path to a junit xml file
# 
# ==== Returns
#
# +hash+ - A hash containing xml objects for the failures, skips, and passes
#
# ==== Examples
#
# load_junit_results("~/junit/latest/beaker_junit.xml")
def load_junit_results(junit_file)
  junit_doc = Nokogiri::XML(File.read(junit_file))

  failures = junit_doc.xpath('//testcase[failure]')
  skips = junit_doc.xpath('//testcase[skip]')
  passes = junit_doc.xpath('//testcase[not(failure) and not(skip)]')

  return {failures: failures, skips: skips, passes: passes}
end


# Extracts the test case id from the test script
#
# ==== Attributes
#
# * +beaker_file+ - Path to a beaker test script
# 
# ==== Returns
#
# +string+ - The test case ID
#
# ==== Examples
#
# testcase_id_from_beaker_script("~/tests/test_the_things.rb") # 1234
def testcase_id_from_beaker_script(beaker_file)
  # Find first matching line
  match = File.readlines(beaker_file).map { |line| line.match(TESTCASE_ID_REGEX) }.compact.first

  match[:testrun_id]
end


# Calculates the path to a beaker test file by combining the junit file path
# with the test name from the junit results.
# Makes the assumption that junit folder that beaker creates will always be 
# 2 directories up from the beaker script base directory.
# TODO somewhat hacky, maybe a config/command line option
#
# ==== Attributes
#
# * +junit_file_path+ - Path to a junit xml file
# * +junit_result+ - Path to a junit xml file
# 
# ==== Returns
#
# +string+ - The path to the beaker script from the junit test result
#
# ==== Examples
#
# load_junit_results("~/junit/latest/beaker_junit.xml")
def beaker_test_path(junit_file_path, junit_result)
  beaker_folder_path = junit_result[:classname]
  test_filename = junit_result[:name]

  File.join(File.dirname(junit_file_path), "../../", beaker_folder_path, test_filename)
end


if __FILE__ == $PROGRAM_NAME
  main parse_opts
end
